### PR TITLE
⚠️ KubeadmControlPlane: Support metadata for machines and propagate to all templates

### LINKED ISSUE #4636: KubeadmControlPlane: Support metadata for machines and propagate to all templates
This issue is to track the support of the adding `clusterv1.ObjectMeta` object information to the KCP object.

Currently, we have a single field called `InfrastructureTemplate` under `spec` which gives information to KCP on how to create the infrastructure of machines https://github.com/kubernetes-sigs/cluster-api/blob/06f58a66353d5a22a2849f2f516ee71e4664c323/controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_types.go#L56-L58

For this issue, we propose a breaking api change to streamline the machine template and align a bit with MachineDeployments as well. The API would look as follows:

```go
type KubeadmControlPlaneSpec struct {
// [...]
	MachineTemplate KubeadmControlPlaneMachineTemplate `json:"machineTemplate"`
// [...]
}

type KubeadmControlPlaneMachineTemplate struct {
	// Standard object's metadata.
	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
	// +optional
	ObjectMeta clusterv1.ObjectMeta `json:"metadata,omitempty"`

	// InfrastructureTemplate is a required reference to a custom resource
	// offered by an infrastructure provider.
	Infrastructure corev1.ObjectReference `json:"infrastructure"`
}
```

The conversion webhooks are going to take care of the following:
- Restore the ObjectMeta information which is not going to be available in v1alpha3.
- Convert the ObjectReference from `spec.infrastructureTemplate` to `spec.machineTemplate.infrastructure` (and vice versa).

Moreover, the `ObjectMeta` information is going to be passed to all templated objects, such as:
- Machine
- InfrastructureMachine (unstructured, dependent on the object reference)
- KubeadmConfig

/kind release-blocking
/kind api-change
/area api

/milestone v0.4

### PR DESCRIPTION
**Pull request by vincepri**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4636

/milestone v0.4
/kind release-blocking
/kind api-change

### CODE REVIEW

File: controlplane/kubeadm/api/v1alpha4/kubeadm_control_plane_types.go [hunk 1]
```diff
@@ -85,6 +84,17 @@ type KubeadmControlPlaneSpec struct {
 	RolloutStrategy *RolloutStrategy `json:"rolloutStrategy,omitempty"`
 }
 
+type KubeadmControlPlaneMachineTemplate struct {
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	ObjectMeta clusterv1.ObjectMeta `json:"metadata,omitempty"`
+
+	// InfrastructureTemplate is a required reference to a custom resource
```
Comment (CecileRobertMichon) [inline controlplane/kubeadm/api/v1alpha4/kubeadm_control_plane_types.go:93]: Infrastructure ?

File: test/e2e/data/infrastructure-docker/v1alpha3/bases/cluster-with-kcp.yaml [hunk 2]
```diff
@@ -52,10 +52,11 @@ metadata:
...
   replicas: ${CONTROL_PLANE_MACHINE_COUNT}
-  infrastructureTemplate:
-    kind: DockerMachineTemplate
-    apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
-    name: "${CLUSTER_NAME}-control-plane"
+  machineTemplate:
+    infrastructureRef:
+      kind: DockerMachineTemplate
+      apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+      name: "${CLUSTER_NAME}-control-plane"
```
Comment (fabriziopandini) [inline test/e2e/data/infrastructure-docker/v1alpha3/bases/cluster-with-kcp.yaml:59]: I don't think v1alpha3 templates should be changed

**Review by fabriziopandini — COMMENTED** [review]
two nits, otherwise lgtm from my side

File: controlplane/kubeadm/internal/filters_test.go [hunk 3]
```diff
@@ -869,6 +957,101 @@ func TestMatchesTemplateClonedFrom(t *testing.T) {
...
+						},
+						Labels: map[string]string{
+							"test": "labels",
+						},
+					},
+					InfrastructureRef: corev1.ObjectReference{
+						Kind:       "GenericMachineTemplate",
+						Namespace:  "default",
+						Name:       "infra-foo",
+						APIVersion: "generic.io/v1",
```
[the remark this answers is not in the document]
Comment (vincepri) [reply controlplane/kubeadm/internal/filters_test.go:980]: Done

File: test/e2e/data/infrastructure-docker/v1alpha3/bases/cluster-with-kcp.yaml (hunk 2 above)
Comment (vincepri) [reply test/e2e/data/infrastructure-docker/v1alpha3/bases/cluster-with-kcp.yaml]: Reverted, good catch

**Comment by fabriziopandini** [comment]
lgtm pending CI signal

**Comment by vincepri** [comment]
These tests are passing locally, even if I run them multiple times — makes me think we're hitting once again some resource starvation now that conversion webhooks are actually enabled

**Comment by fabriziopandini** [comment]
@vincepri before moving on with this PR I'm wondering if we should add `spec.machineTemplate.Metadata` and  `spec.machineTemplate.infrastructureRef` in the ControlPlane contract so we can assume those field exists for all the controlPlane providers.

**Comment by vincepri** [comment]
@fabriziopandini Is that for the ClusterClass proposal? If so, I'm ok waiting until the proposal PRs this repository and changes the contract to support ClusterClass. cc @srm09, we could also do that separately now, although I want to make sure to track the work

**Comment by fabriziopandini** [comment]
Yes, I was thinking to ClusterClass; I'm adding comments to the proposal doc.
/lgtm

**Pull request merged**